drivers/pinmux: stm32: API rework

erwango opened the pull request:
- Revisit stm32_dt_pincltrl into one single call: stm32_dt_pinctrl_configure
- F1: Rather than spreading it in various impacted drivers, centralize remapping function in stm32 pinmux driver.

Based on #29231.

erwango committed:
drivers/pinctrl: stm32f1: Provide a function which centralize remap

Centralize AFIO remapping into one single function.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
drivers/pinmux: stm32: Get remap functions visible across family

Move pinctrl remap functions out of stm32f1 definition in order
to get it available to all series.
Allows use of more IS_ENABLED macros in calling drivers and make
code more readable.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

galak reviewed (commented):
Can you move

```
#if DT_HAS_COMPAT_STATUS_OKAY(st_stm32f1_pinctrl)
		int remap;
		/* Check that remap configuration is coherent across pins */
		remap = stm32_dt_pinctrl_remap_check(cfg->pinctrl_list,
						     cfg->pinctrl_list_size);
		if (remap < 0) {
			return remap;
		}

		stm32_dt_pinctrl_remap_set((uint32_t)cfg->i2c, remap);
#endif /* DT_HAS_COMPAT_STATUS_OKAY(st_stm32f1_pinctrl) */
```

into `stm32_dt_pinctrl_configure`?  So each driver doesn't have to do that?

erwango committed:
drivers/pinmux: stm32: Provide unique API to stm32_dt_pinctrl

Set stm32_dt_pinctrl_configure function as the unique entry point
to STM32 DT pinctrl management.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango commented:
> Can you move [...]

@galak Went one step further and merge _check and _set functions, as there is no value doing this inside drivers.
Though, I kept:
- the `if STM32F1` condition since `remap` has no meaning on other series
- the capacity to detect and report errors so wrong configuration doesn't go undetected.

gmarull commented on drivers/i2c/i2c_ll_stm32.c:
```diff
@@ -189,31 +189,17 @@ static int i2c_stm32_init(const struct device *dev)
+		if (IS_ENABLED(DT_HAS_COMPAT_STATUS_OKAY(st_stm32f1_pinctrl))) {
+			int remap;
+			/* Check remap configuration is coherent across pins */
+			remap = stm32_dt_pinctrl_remap(cfg->pinctrl_list,
```
Wouldn't it be possible to even embed this part inside `stm32_dt_pinctrl_configure`?

gmarull commented on drivers/i2c/i2c_ll_stm32.c:
```diff
@@ -189,31 +189,17 @@ static int i2c_stm32_init(const struct device *dev)
 #endif
 
 	if (cfg->pinctrl_list_size != 0) {
```
In fact, the `0` size check is not needed, function should just return if given an empty array of pins.

gmarull reviewed (commented):
Thanks for this change, driver code will be much more readable. Left a couple of comments.

erwango commented:
@gmarull Txs for review, indeed there is room to push this further.

alexanderwachter reviewed (approved):
I don't have a STM32F1 to test, but I like The general idea. +1 from me.

martinjaeger commented on boards/arm/disco_l475_iot1/disco_l475_iot1.dts:
```diff
@@ -121,9 +121,13 @@
 	status = "okay";
 };
 
+&spi3_miso_pc11{ slew-rate = "very-high-speed";};
```
```suggestion
&spi3_miso_pc11 {
        slew-rate = "very-high-speed";
};
```

erwango replied:
@martinjaeger this is actually coming from https://github.com/zephyrproject-rtos/zephyr/pull/29231 on which current PR is based on (besides, if you don't like this notation, you'll find more to hint on https://github.com/zephyrproject-rtos/zephyr/pull/29267).
Based on this can you reconsider your review on current PR and continue on #29231 or #29267 ?

martinjaeger replied:
Sure, will have a look at the other PRs aswell.

martinjaeger reviewed (changes requested):
Great improvement to clean up the driver-specific code. Only got a tiny style comment.

Don't have an F1 board for testing either, but went through all the commits and the changes look good to me.

erwango committed:
drivers/pinmux: stm32: remap code should be guarded under node status

CAN remaps were guarded by CMSIS defines which are always defined
for a given SoC.
Though under this control, we're using DT_ macros that expect
a certain node to be available, which otherwise leads to cmopilation
issue.
Align CAN node remap code on other peripherals code and check for
node availability.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
drivers/pinmux: stm32f1: Revise timers possible remap values

Only pinctrl for tim2 can have a REMAP_FULL value.
Usual value is REMAP_1.
Fix this

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

galak commented:
@gmarull can you take another look.

gmarull reviewed (approved):
Great work!
